services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/service/impl/DpmServiceImpl.java

```diff
@@ -144,20 +153,92 @@ private UnicastGoalState buildUnicastGoalState(NetworkConfiguration networkConfi
                     networkConfig, pulsarHostPortEntities, pulsarDataPlaneClient));
         }
 
-        localCache.addSubnetPorts(networkConfig);
+        localCache.updateLocalCache(networkConfig);
 
         return statusList;
     }
 
-    private List<String> createNeighborConfiguration(NetworkConfiguration networkConfig) throws Exception {
-        return null;
+    private List<String> processNeighborConfiguration(NetworkConfiguration networkConfig) throws Exception {
+        Map<String, NeighborInfo> neighborInfos = networkConfig.getNeighborInfos();
+        Map<String, List<NeighborEntry>> neighborTable = networkConfig.getNeighborTable();
+        List<UnicastGoalState> unicastGoalStates = new ArrayList<>();
+        MulticastGoalState multicastGoalState = new MulticastGoalState();
+
+        if (neighborTable == null || neighborInfos == null) {
+            throw new NeighborInfoNotFound();
+        }
+
+        Map<String, List<NeighborInfo>> hostNeighbors = new HashMap<>();
+        for (Map.Entry<String, List<NeighborEntry>> entry: neighborTable.entrySet()) {
+            String portIp = entry.getKey();
+            NeighborInfo localInfo = neighborInfos.get(portIp);
+            if (localInfo == null) {
+                throw new NeighborInfoNotFound();
+            }
+
+            String hostIp = localInfo.getHostIp();
+            if (!hostNeighbors.containsKey(hostIp)) {
+                hostNeighbors.put(hostIp, new ArrayList<>());
+            }
+
+            List<NeighborEntry> neighborEntries = entry.getValue();
+            for (NeighborEntry neighborEntry: neighborEntries) {
+                String neighborIp = neighborEntry.getNeighborIp();
+                NeighborInfo neighborInfo = neighborInfos.get(neighborIp);
+                if (neighborInfo == null) {
+                    throw new NeighborInfoNotFound();
+                }
+
+                hostNeighbors.get(hostIp).add(neighborInfo);
+                multicastGoalState.getHostIps().add(neighborInfo.getHostIp());
+            }
+
+            //Add neighborInfo to multicastGoalState
+            Neighbor.NeighborState neighborState = neighborService.buildNeighborState(
+                    NeighborType.L3, localInfo, networkConfig.getOpType());
+            multicastGoalState.getGoalStateBuilder().addNeighborStates(neighborState);
+        }
+
+        for (Map.Entry<String, List<NeighborInfo>> entry: hostNeighbors.entrySet()) {
+            String hostIp = entry.getKey();
+            List<NeighborInfo> hostNeighborInfos = entry.getValue();
+
+            /**
+             * At present, there are only L3 neighbors in the neighbor table,
+             * and the processing of L2 neighbors should be considered in the future.
+             */
```

Comment:
Can you elaborate a bit here? Neighbor table also includes L2 neighbors if my memory is right :-)

Reply:
In this case is for update-l3-neighbor, so there will be no L2 neighbors here.

Reply:
aha sounds good.